### DISCUSSION TITLE
Categorizing types of things in Janet? (plus cfunctions and abstract types)

### DISCUSSION THREAD

**Discussion by uvtc** [janet-lang/janet#556] [General]
What are the types of things in Janet called? As in, you have simple objects like `1`, `5.2`, `true`, `:this-and-that`, and symbols, but also compound objects like `@[1 2]`. Maybe the latter are just referred to as "data structures"? But what are the former called?

In chat it was pointed out to me that there's an `idempotent?` function that seems to suggest maybe the term "atomic" for these simple objects?

It was also pointed out to me by @sogaiu that the types of things in Janet are listed right at <https://github.com/janet-lang/janet/blob/master/src/include/janet.h#L399-L416>.

The discussion began getting deeper, touching on `function?`, `cfunction?`, and abstract types.

Are cfunctions functions in Janet that are implemented in C? Which ones are these?

What are abstract types in Janet?

**Comment by sogaiu** [comment]
Here are some possibly relevant bits:

* [C functions](https://janet-lang.org/docs/abstract_machine.html#C-functions)
* [Abstract types](https://janet-lang.org/docs/object_oriented.html)

As to whether something is a cfunction, I think one may be able to use `cfunction?` to tell.

Also, the `doc` macro can provide some of that info.  For example:
```
repl:1:> (doc os/time)


    cfunction

    (os/time)

    Get the current time expressed as the number of seconds since January
    1, 1970, the Unix epoch. Returns a real number.


nil
```

**Comment by yumaikas** [comment]
A cfunction is a binding from Janet to a function implemented in C. Most interpreted languages have that sort of construct, it varies in if they expose that in the type system or not. JavaScript on Chrome used to indicate if a function was "native" or composed via JS

I can't speak for Janet abstract types in detail, but I saw a comment about them being similar to a Lua userdata.

As far as categories of types, I'd suggest that you could break types in Janet into values and containers. Values would be integers, books, and symbols. Containers would be more like tables, structs, arrays, tuples and buffers. A string is the hybrid between the two, containing bytes, but usually considered as a value, and immutable. 
Functions I would also classify as values.

But, for now, bakpakin is the ultimate authority on this. I do like that every container type has a mutable and an immutable version. It provides a nice symmetry.

**Comment by MikeBeller** [comment]
I did notice that maybe there aren't clear 'categories' of types in Janet.   For what it's worth:  Common LISP has "ATOM'   to distinguish between atomic and compound (list) data.

Perhaps there should be a function `atom?` which would be true for number, boolean, nil, symbol.  But how should it work for int/u64 which is an abstract type?  (Maybe if janet_abstract->next is null then `atom?` should return true?)

And on this same topic -- anyone notice that Janet's `indexed?` returns true for array/tuple/struct/table, but false for string and buffer?  But I think the indexed combinators work on strings and buffer.  (Note: it is true that `indexed?` matches its documentation.)

Mike

**Reply by bakpakin** [reply]
Yes, there are some fuzzy boundaries for combining different types, but the difference between arrays/tuples and strings/symbols/keywords/buffers is that the former are recursive types - they can contain other recursive types. The latter contain just sequences of numbers.

Perhaps different predicates could make this clearer.

**Comment by uvtc** [comment]
I see that `idempotent?` (thanks, @ahungry ) returns `true` for both `os/time` (a cfunction) and for my own `foo` function:

~~~
(defn foo [x] (* x x))
(idempotent? os/time) #=> true
(idempotent? foo)     #=> true
(idempotent? 1)       #=> true
(idempotent? :hey)    #=> true
(idempotent? 'foo)    #=> true
~~~

I'd always thought of "idempotent" to mean a "pure" (side-effect free) function.

**Reply by pepe** [reply]
To clarify, what is the side-effect of the os/time? I see it as idempotent.

**Reply by uvtc** [reply]
Sorry for any confusion. That was just my own understanding of what the term means.

~~~
#!/usr/bin/env janet

(var x 2)

(defn foo
  []
  (print "side-effecting!")
  (set x 3))

(print x)
(foo) # 2
(print (idempotent? foo)) #=> true
(print x) # 3
~~~

**Reply by bakpakin** [reply]
idemptotent? is meant to mean "if this were compiled as the body of the function, would that function be idempotent?" This is a useful property for writing macros, and as long as we are conservative lets us not create extra bindings for things that are constant.

You can see how idempotent? is used in boot.janet for some macros.

**Reply by uvtc** [reply]
What specifically do you mean by, "..., and as long as we are conservative lets us not create extra bindings for things that are constant"?

**Comment by uvtc** [comment]
Ok, I think then the terms I was looking for were recursive (tuples, arrays, structs, tables) vs non-recursive (possibly aka "atomic") (symbols, keywords, strings, buffers, **edit:** booleans too).

And it's interesting to see the other categorizations:

  * function vs cfunction
  * Janet type vs abstract type
  * `indexed?` vs `dictionary?` vs `bytes?`
  * mutable vs immutable

**Reply by sogaiu** [reply]
Note that according to: https://github.com/janet-lang/janet/blob/2abb87eb631d11910760d790efaef19c2604b671/src/include/janet.h#L398-L416 `JANET_ABSTRACT` is a member(?) of `JanetType`.
